src/audio/dai-zephyr.c

```diff
@@ -846,16 +692,81 @@ static int dai_params(struct comp_dev *dev, struct sof_ipc_stream_params *params
 		 * component is able to convert stream with different
 		 * frame_fmt's (using pcm converter).
 		 */
-		hw_params.frame_fmt = dev->ipc_config.frame_fmt;
+		hw_params->frame_fmt = dev->ipc_config.frame_fmt;
 		buffer_c = buffer_acquire(dd->dma_buffer);
-		buffer_set_params(buffer_c, &hw_params,
-				  BUFFER_UPDATE_FORCE);
+		buffer_set_params(buffer_c, hw_params, BUFFER_UPDATE_FORCE);
 		buffer_release(buffer_c);
 	}
 
-	return dev->direction == SOF_IPC_STREAM_PLAYBACK ?
-		dai_playback_params(dev, period_bytes, period_count) :
-		dai_capture_params(dev, period_bytes, period_count);
+	return 0;
+}
+
+static int dai_params(struct comp_dev *dev, struct sof_ipc_stream_params *params)
+{
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct dma_sg_config *config = &dd->config;
+	uint32_t period_bytes = 0;
+	uint32_t period_count = 0;
+	int err = 0;
+
+	comp_dbg(dev, "dai_params()");
+
+	/* check if already configured */
+	if (dev->state == COMP_STATE_PREPARE) {
+		comp_info(dev, "dai_set_dma_buffer() component has been already configured.");
+		return 0;
+	}
+
+	/* can set params on only init state */
+	if (dev->state != COMP_STATE_READY) {
+		comp_err(dev, "dai_set_dma_buffer(): comp state %d, expected COMP_STATE_READY.",
+			 dev->state);
+		return -EINVAL;
+	}
```

Comment:
now, this time I'm not dreaming, am I? Is this intentional - moving the above two checks before calls to `ipc_dai_data_config()` and `dai_verify_params()`?